### DISCUSSION TITLE
What am i doing wrong here? Wrapper not working...

### DISCUSSION THREAD

**Discussion by rica-carv** [e107inc/e107#5326] [Q&A Plugins] [answered]
I've searched everywhere, even copied code form other plugins, and my wrapper on a menu simply doesn't works... It's getting me crazy....

What am i doing wrong in my code?

main_menu.php file code:
```
$template = e107::getTemplate('eforum','eforum_menu', 'foruminfo', true, true);

$sc = e107::getScBatch("foruminfo", "eforum");
$sc->wrapper('eforum_menu/foruminfo');

$foruminfo = e107::getParser()->parseTemplate($template['foruminfo'], true, $sc);
```
eforum_menu_template.php file code:
```
$EFORUM_MENU_TEMPLATE['foruminfo'] = "<div class='d-flex col justify-content-center align-items-center'>
{EFIM_ICONKEY}
</div><br>
<div class='d-flex col justify-content-center align-items-center'>
{EFIM_USERINFOX}
</div>";

$EFORUM_MENU_WRAPPER['foruminfo']['EFIM_ICONKEY']		= 	"<div class='card col'>{---}</div>";
```

foruminfo_shortcode.php file code:
```
class plugin_eforum_foruminfo_shortcodes extends e_shortcode
{
function __construct()
{....}
function sc_efim_iconkey()
{	return "ICONKEY";}
function sc_efim_userinfox()
{	return "USERINFOX";}
}
```

Shortcodes is loaded, main template also and renders, but no wrapper....

Can anyone share some thoughts about this weird issue?????

**Comment by Vodhin** [comment]
Looks like the code is incomplete?  $tp->parseTemplate('{MY_TEMPLATE}', false, $sc));

**Reply by rica-carv** [reply]
@Vodhin Updated, but the error is not in template parsing, is in `$sc->wrapper('eforum_menu/foruminfo');`, i think....

**Reply by Jimmi08** [reply]
It sounds too much familiar. Underscore in plugin name?  I hit similar issue years ago. 
Can you share your plugin?  In private repo if you want it this way.

**Reply by rica-carv** [reply]
@Jimmi08 I don't get it, simply tryed everything, and nothing solves the issue.
If it's any help, [my repo is here](https://github.com/rica-carv/eforum). Keep in mind it's a work in progress, so it's a little bit buggy....
And thanks for helping. 👍

**Reply by Jimmi08** [reply]
@rica-carv thanks By the way, the forum plugin in my use is here:  https://github.com/Jimako-e107-plugins/forum

**Reply by Jimmi08** [reply]
If you switch to PHP 8, you would get errors:
Fatal error: Uncaught TypeError: sizeof(): Argument # 1 ($value) must be of type Countable|array, null given in C:\wamp64\www\e107core\e107_plugins\eforum\foruminfo_menu.php on line 590
Easy fix: add at the start $options = array();

Fatal error: Uncaught Error: Attempt to assign property "forum_sc" on null in \shortcode_handler.php on line 1888
$sc = e107::getScBatch("foruminfo", "eforum"); -  so your wrapper fails, but you don't see it in PHP 7

this fails:   $this->forum_sc = e107::getScBatch('forum', true); 

My advice: don't use them, copy them.  Yeah, I know, duplicate code...  but some things in e107 work only of luck, or special cases.  If you want to use them more advanced way, they fail. I learned this with the download or news plugin.   It is always the issue with legacy plugins. 

Or your shortcode should be in e_shortcode, not in batch...   
Sorry, I can't help more

**Reply by rica-carv** [reply]
> @rica-carv thanks By the way, the forum plugin in my use is here: https://github.com/Jimako-e107-plugins/forum

@Jimmi08 Thanks, but you changed the plugin itself, as far i see, right?
And thanks for the help. I'm indeed using php 7.4 in my development server. I do that because free hosting normally doesn't have the last php version available, so i go with the bare bones minimum php....

The funny thing is that in PHP 8 it dosn't work, but with 7.4 it works, because i got that `$this->forum_sc = e107::getScBatch('forum', true);` working....
And e_shortcodes is not an option, since i don't want that shortcodes availabe in the entire site, just in forum pages....
Since i'm a little stubborn, i'll still try to find this bug in php 7.4.
Anyway thanks for the help

**Reply by Jimmi08** [reply]
PHP 7.4 just hid errors and the code continuea, but you have unexpectable results - like your wrapper is not working. 
But you should see at least warning in PHP logs (but e107 hids a lot of it)

Yes, I don't use the core forum plugin anymore. I was in a similar position as you some years ago... I tried to fix it, I created many issues... nobody cared...   Probably, my PRs weren't well coded so...    
I don't understand why if they are not fixing core plugins (f.e forum attachments) it is still in core... 
Long story, I solved it already my way...  At that time they could test everything on my e107sk forum, no interest...   So, what to say?
I use e107 if I need a custom application but as cms.  Pages and news are nightmares. 
Don't take me wrong, it is awesome code and people did great work... e107 handlers are the best to use. But I need to make money from clients and I can't offer them not fully working cms.   So I use e107 as PHP framework, with custom plugins, nothing more.   

By the way, I tried it at first with plugin forum_extended (like your eforum) but it failed exactly on shortcodes...   
If you notice, forum batch shortcodes use different class name as standard plugin batch shortcodes...

**Reply by rica-carv** [reply, marked as answer]
Nailed it!!!!!!!!!!!

Just got to use `$SC_WRAPPER['EFIM_ICONKEY'] = "<div class='card col'>{---}</div>";`

What a headache!!!!

> If you notice, forum batch shortcodes use different class name as standard plugin batch shortcodes...

@Jimmi08 You're right, even the shortcodes locations are different, so what would be standard locations and standard names for shortcodes classes?
Plugin root or plugin/shortcodes/batch ???
`class plugin_<plugin name>_<shortcode file name>_shortcodes` or `class <shortcode file name>_shortcodes` ???

I'm quite lost in space here....

Vary mixed code, and once more, with still support legacy old code, if it won't work anymore??? Anyway, to make old plugins work, they must be forced anyway to be upgraded to new PHP code!!! I simply don't get it....

**Reply by Jimmi08** [reply]
Exactly my point - old globals work - $SC_WRAPPER, but OOP stuff doesn't.     
class plugin_<plugin name>_<shortcode file name>_shortcodes - this for batch
class <shortcode file name>_shortcodes for shortcodes in root
+ e_shortcode. 

I prefer batch because you can have more files of shortcodes...  like with templates.
